src/parser/types.rs

```diff
@@ -193,38 +193,34 @@ pub enum TyKind {
 }
 
 impl TyKind {
-    pub fn match_expected(&self, expected: &TyKind) -> bool {
-        match (self, expected) {
-            (TyKind::Field { .. }, TyKind::Field { .. }) => true,
-            (TyKind::Array(lhs, lhs_size), TyKind::Array(rhs, rhs_size)) => {
-                lhs_size == rhs_size && lhs.match_expected(rhs)
-            }
-            (
-                TyKind::Custom { module, name },
-                TyKind::Custom {
-                    module: expected_module,
-                    name: expected_name,
-                },
-            ) => module == expected_module && name == expected_name,
-            (x, y) if x == y => true,
-            _ => false,
-        }
-    }
-
-    pub fn same_as(&self, other: &TyKind) -> bool {
+    pub fn compare_with(&self, other: &TyKind, exact_match: bool) -> bool {
         match (self, other) {
+            // Handle the Field types with optional exact match logic
+            (TyKind::Field { constant: true }, TyKind::Field { constant: false })
+            | (TyKind::Field { constant: false }, TyKind::Field { constant: true }) => {
+                !exact_match
+            }
+            (TyKind::Field { .. }, TyKind::Field { .. }) => true,
```

Comment:
so I think this is the tricky part, I actually think we should write this:

```rust
(TyKind::Field { constant: observed }, TyKind::Field { constant: expected }) =>
  (!expected || observed) 
```

and this assume that the expected/signature is the `other` and the one we're checking against it is `self` which is not necessarily implied by the name of the function. So perhaps we should keep `match_expected` as a name